model/jpa/src/main/java/org/keycloak/models/jpa/session/PersistentUserSessionEntity.java

```diff
@@ -47,14 +47,14 @@
                 " AND sess.realmId = :realmId AND sess.userId = :userId ORDER BY sess.userSessionId"),
         @NamedQuery(name="findUserSessionsByClientId", query="SELECT sess FROM PersistentUserSessionEntity sess INNER JOIN PersistentClientSessionEntity clientSess " +
                 " ON sess.userSessionId = clientSess.userSessionId AND clientSess.clientId = :clientId WHERE sess.offline = :offline " +
-                " AND sess.userSessionId = clientSess.userSessionId AND sess.realmId = :realmId ORDER BY sess.userSessionId"),
-        @NamedQuery(name="findUserSessionsCountsByClientId", query="SELECT clientSess.clientId, count(clientSess) " +
-                " FROM PersistentUserSessionEntity sess INNER JOIN PersistentClientSessionEntity clientSess " +
-                " ON sess.userSessionId = clientSess.userSessionId " +
-                // find all available offline user-session for all clients in a realm
-                " WHERE sess.offline = :offline " +
-                " AND sess.userSessionId = clientSess.userSessionId AND sess.realmId = :realmId " +
-                " GROUP BY clientSess.clientId")
+                " AND sess.realmId = :realmId ORDER BY sess.userSessionId"),
+        @NamedQuery(name="findUserSessionsByExternalClientId", query="SELECT sess FROM PersistentUserSessionEntity sess INNER JOIN PersistentClientSessionEntity clientSess " +
+                " ON sess.userSessionId = clientSess.userSessionId AND clientSess.clientStorageProvider = :clientStorageProvider AND clientSess.externalClientId = :externalClientId WHERE sess.offline = :offline " +
+                " AND sess.realmId = :realmId ORDER BY sess.userSessionId"),
+        @NamedQuery(name="findClientSessionsClientIds", query="SELECT clientSess.clientId, clientSess.externalClientId, clientSess.clientStorageProvider" +
+                " FROM PersistentClientSessionEntity clientSess INNER JOIN PersistentUserSessionEntity sess ON clientSess.userSessionId = sess.userSessionId " +
+                " WHERE sess.offline = :offline AND sess.realmId = :realmId " +
+                " GROUP BY clientSess.clientId, clientSess.externalClientId, clientSess.clientStorageProvider")
```

Comment:
I think your point is correct that we might lose some entries. I removed the `GROUP BY` but I want to recheck that all databases are OK with that. (Can't do it ATM because keycloak-build-pipeline is not working). If some database will complain about missing GROUP BY I can still select the whole `PersistentClientSessionEntity` and then process it in Java code.